### PR TITLE
[BUGFIX] Ensure that GX supports database column names in lowercase and mixcase for Oracle, DB2, and Snowflake

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
For certain database management systems (DBMSs), such as Oracle, DB2, and Snowflake, standard casing is in all capitals.  However, SQLAlchemy standardizes on all lowercase.  This change fixes issues with Metrics part of GX Framework, whereby quoting of lowercase table names is provided so that SQLAlchemy can pass them to DBMS and access table properly.

### Remarks
* JIRA: DX-307
* For clarity, this pull request should be reviewed after pull requests #8302 and #8327 have been merged.

- [ ] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted

    ```
    black .

    ruff . --fix
    ```
- [x] Appropriate tests and docs have been updated

For more details, see our [Contribution Checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist), [Coding style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style), and [Documentation style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/docs_style).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: great_expectations/datasource/fluent/sql_datasource.py [hunk 1]
```diff
@@ -802,7 +802,7 @@ class TableAsset(_SQLAsset):
 
     # Instance fields
     type: Literal["table"] = "table"
-    table_name: str = pydantic.Field(
+    table_name: Union[str] = pydantic.Field(
```
Comment (cdkini) [inline great_expectations/datasource/fluent/sql_datasource.py]: Union between what? I don't think this type is correct

File: great_expectations/expectations/metrics/util.py [hunk 2]
```diff
@@ -606,6 +606,55 @@ def column_reflection_fallback(
...
+    column name in "Batch" column list is quoted, then corresponding column name in "metric_domain_kwargs" is also quoted.
+
+    Args:
+        metric_domain_kwargs: Original "metric_domain_kwargs" dictionary of attribute key-value pairs.
+        batch_columns_list: Actual "Batch" column list (e.g., output of "table.columns" metric).
+
+    Returns:
+        metric_domain_kwargs: Updated "metric_domain_kwargs" dictionary with quoted column names, where appropriate.
+    """
+    if "column" in metric_domain_kwargs:
```
Comment (cdkini) [inline great_expectations/expectations/metrics/util.py:625]: Seems like we know the possible keys for `metric_domain_kwargs` - could we use a TypedDict? You'll find other examples if you grep the codebase

File: tests/expectations/metrics/test_metrics_util.py [hunk 3]
```diff
@@ -272,3 +274,149 @@ def test_get_unexpected_indices_for_multiple_pandas_named_indices_named_unexpect
...
+        batch_columns_list=test_column_names,
+    )
+    assert metric_domain_kwargs["column"] == quoted_column_name
+
+    metric_domain_kwargs = get_dbms_compatible_metric_domain_kwargs(
+        metric_domain_kwargs={"column": "travel_agents"},
+        batch_columns_list=test_column_names,
+    )
+    assert metric_domain_kwargs["column"] == quoted_column_name
+    assert metric_domain_kwargs["column"] != "TRAVEL_AGENTS"
```
Comment (cdkini) [inline tests/expectations/metrics/test_metrics_util.py]: Feels like we're testing three different cases here - would it make sense to use `pytest.mark.parameterize`?

File: tests/expectations/metrics/test_metrics_util.py [hunk 4]
```diff
@@ -272,3 +274,149 @@ def test_get_unexpected_indices_for_multiple_pandas_named_indices_named_unexpect
...
+            "column_list": ["ACTORS", "travel_agents", "FarmAnimals", "Household_Pets"]
+        },
+        batch_columns_list=test_column_names,
+    )
+    assert metric_domain_kwargs["column_list"] == [
+        "ACTORS",
+        quoted_column_name_0,
+        quoted_column_name_1,
+        quoted_column_name_2,
+    ]
```
Comment (cdkini) [inline tests/expectations/metrics/test_metrics_util.py]: Any possible issues with sorting here? Are we guaranteed order?

File: great_expectations/datasource/fluent/sql_datasource.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/datasource/fluent/sql_datasource.py]: @cdkini Looks like an old typo.  Fixed/committed.  Thank you.

File: tests/expectations/metrics/test_metrics_util.py (hunk 3 above)
Comment (alexsherstinsky) [reply tests/expectations/metrics/test_metrics_util.py]: @cdkini I can; the tests are not exactly the same and there are not too many of them.  Do you think it is worth parametrizing?  Please let me know what you think.  Thanks.

File: tests/expectations/metrics/test_metrics_util.py (hunk 4 above)
Comment (alexsherstinsky) [reply tests/expectations/metrics/test_metrics_util.py]: @cdkini I have not encountered them; the method under test does not change the order.  I will have to see what happens in the CI tests to be fully sure (locally these tests passed).  Thanks.

File: great_expectations/expectations/metrics/util.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/metrics/util.py:625]: @cdkini `metric_domain_kwargs` permeates the codebase, so I would prefer this improvement to be a separate JIRA ticket (impacting `MetricConfiguration`, `ValidationGraph`, `MetricsCalculator`, and many more core classes).  Happy to discuss.  Thanks.

File: great_expectations/expectations/metrics/map_metric_provider/map_condition_auxilliary_methods.py [hunk 5]
```diff
@@ -109,26 +109,21 @@ def _pandas_map_condition_index(
...
                 domain_records_df[column_name].notnull()
             ]
-        domain_column_name_list.append(column_name)
 
-    # multi-column map expectations
-    elif "column_list" in accessor_domain_kwargs:
-        column_list: List[Union[str, sqlalchemy.quoted_name]] = accessor_domain_kwargs[
-            "column_list"
-        ]
-        verify_column_names_exist(
```
Comment (anthonyburdi) [inline great_expectations/expectations/metrics/map_metric_provider/map_condition_auxilliary_methods.py:119]: This looks like a behavior change (deleting the verification that column names exist), should this maybe be in a separate PR?

File: great_expectations/expectations/metrics/map_metric_provider/map_condition_auxilliary_methods.py [hunk 6]
```diff
@@ -198,14 +200,6 @@ def _pandas_map_condition_query(
                 domain_records_df[column_name].notnull()
             ]
 
-    elif "column_list" in accessor_domain_kwargs:
-        column_list: List[Union[str, sqlalchemy.quoted_name]] = accessor_domain_kwargs[
-            "column_list"
-        ]
-        verify_column_names_exist(
```
Comment (anthonyburdi) [inline great_expectations/expectations/metrics/map_metric_provider/map_condition_auxilliary_methods.py:205]: This looks like a behavior change (deleting the verification that column names exist), should this maybe be in a separate PR?

File: tests/datasource/fluent/test_sql_datasources.py [hunk 7]
```diff
@@ -28,5 +31,143 @@ def test_kwargs_are_passed_to_create_engine(mocker: MockerFixture):
...
+
+
+@pytest.mark.unit
+def test_table_quoted_name_type_all_upper_case_normalizion_is_noop():
+    """
+    DBMS entity names (table, column, etc.) must adhere to correct case insensitivity standards.  All upper case is
+    standard for Oracle, DB2, and Snowflake, while all lowercase is standard for SQLAlchemy; hence, proper conversion to
+    quoted names must occur.  This test ensures that all upper case entity usage does not undergo any conversion.
+    """
+    table_names_in_dbms_schema: list[str] = [
```
Comment (anthonyburdi) [inline tests/datasource/fluent/test_sql_datasources.py]: Just a small item: We only need one or maybe two (with, without underscore) names for this test, no need to test the whole list.

File: tests/datasource/fluent/test_sql_datasources.py [hunk 8]
```diff
@@ -28,5 +31,143 @@ def test_kwargs_are_passed_to_create_engine(mocker: MockerFixture):
...
+
+
+@pytest.mark.unit
+def test_table_quoted_name_type_all_lower_case_normalizion_full():
+    """
+    DBMS entity names (table, column, etc.) must adhere to correct case insensitivity standards.  All upper case is
+    standard for Oracle, DB2, and Snowflake, while all lowercase is standard for SQLAlchemy; hence, proper conversion to
+    quoted names must occur.  This test ensures that all lower case entity usage undergo conversion to quoted literals.
+    """
+    table_names_in_dbms_schema: list[str] = [
```
Comment (anthonyburdi) [inline tests/datasource/fluent/test_sql_datasources.py]: Same comment about size of list.

File: tests/expectations/metrics/test_metrics_util.py (hunk 3 above)
Comment (anthonyburdi) [reply tests/expectations/metrics/test_metrics_util.py]: I agree, I think it would be good to parametrize or split this test up to make it more clear / closer to a single assert.

File: tests/expectations/metrics/test_metrics_util.py [hunk 9]
```diff
@@ -272,3 +274,149 @@ def test_get_unexpected_indices_for_multiple_pandas_named_indices_named_unexpect
...
+        quoted_column_name,
+    ]
+
+    metric_domain_kwargs: dict
+
+    metric_domain_kwargs = get_dbms_compatible_metric_domain_kwargs(
+        metric_domain_kwargs={"column_A": "ACTORS", "column_B": quoted_column_name},
+        batch_columns_list=test_column_names,
+    )
+    assert (
```
Comment (anthonyburdi) [inline tests/expectations/metrics/test_metrics_util.py]: Could we split these asserts instead of using `and`? Just so it's easier to see which fails. This applies to the rest of the tests as well.

File: tests/expectations/metrics/test_metrics_util.py [hunk 10]
```diff
@@ -272,3 +274,149 @@ def test_get_unexpected_indices_for_multiple_pandas_named_indices_named_unexpect
         "Error: The list of domain columns is currently empty. Please check your "
         "configuration."
     )
+
+
+def test_get_dbms_compatible_metric_domain_column_kwargs_column_not_found():
```
Comment (anthonyburdi) [inline tests/expectations/metrics/test_metrics_util.py]: These tests should have markers

File: great_expectations/expectations/metrics/map_metric_provider/map_condition_auxilliary_methods.py (hunk 5 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/metrics/map_metric_provider/map_condition_auxilliary_methods.py:119]: @anthonyburdi the new code verifies, making this here unnecessary (the tests confirmed this).

File: great_expectations/expectations/metrics/map_metric_provider/map_condition_auxilliary_methods.py (hunk 6 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/metrics/map_metric_provider/map_condition_auxilliary_methods.py:205]: @anthonyburdi Please see `https://github.com/great-expectations/great_expectations/pull/8345#discussion_r1267315680` -- we achieved code savings.  Thanks.

File: tests/datasource/fluent/test_sql_datasources.py (hunk 7 above)
Comment (alexsherstinsky) [reply tests/datasource/fluent/test_sql_datasources.py]: @anthonyburdi Should I make it smaller, or leave as is (nice names :-) there)?

File: tests/datasource/fluent/test_sql_datasources.py (hunk 8 above)
Comment (alexsherstinsky) [reply tests/datasource/fluent/test_sql_datasources.py]: @anthonyburdi Should I shorten it?

File: tests/expectations/metrics/test_metrics_util.py (hunk 9 above)
Comment (alexsherstinsky) [reply tests/expectations/metrics/test_metrics_util.py]: @anthonyburdi I think that the `AND` is because they are part of the `MetricDomainTypes.COLUMN_PAIR` domain type.  I see them as key going together -- or am I missing the obvious here?  Thanks.

File: tests/expectations/metrics/test_metrics_util.py (hunk 10 above)
Comment (alexsherstinsky) [reply tests/expectations/metrics/test_metrics_util.py]: @anthonyburdi Good catch -- markers added (committed).  Thanks.

File: tests/expectations/metrics/test_metrics_util.py (hunk 3 above)
Comment (alexsherstinsky) [reply tests/expectations/metrics/test_metrics_util.py]: @cdkini Upon conversation with @anthonyburdi we came up with an elegant way to parametrize -- in next commit.

Comment (alexsherstinsky) [reply tests/expectations/metrics/test_metrics_util.py]: @cdkini this has been committed.

File: tests/expectations/metrics/test_metrics_util.py (hunk 4 above)
Comment (alexsherstinsky) [reply tests/expectations/metrics/test_metrics_util.py]: @cdkini I added sorting -- thanks.

File: great_expectations/expectations/metrics/map_metric_provider/map_condition_auxilliary_methods.py (hunk 5 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/metrics/map_metric_provider/map_condition_auxilliary_methods.py:119]: @anthonyburdi Marking as resolved per our discussion.  Thanks.

File: tests/datasource/fluent/test_sql_datasources.py (hunk 7 above)
Comment (alexsherstinsky) [reply tests/datasource/fluent/test_sql_datasources.py]: @anthonyburdi This has been committed -- thanks.

File: tests/datasource/fluent/test_sql_datasources.py (hunk 8 above)
Comment (alexsherstinsky) [reply tests/datasource/fluent/test_sql_datasources.py]: @anthonyburdi This has been committed -- thanks.

File: tests/expectations/metrics/test_metrics_util.py (hunk 9 above)
Comment (alexsherstinsky) [reply tests/expectations/metrics/test_metrics_util.py]: @anthonyburdi The assertions have been split -- thanks!

File: tests/expectations/metrics/test_metrics_util.py [hunk 11]
```diff
@@ -272,3 +295,149 @@ def test_get_unexpected_indices_for_multiple_pandas_named_indices_named_unexpect
...
+    )
+    quoted_column_name_2: sqlalchemy.quoted_name = sqlalchemy.quoted_name(
+        value="Household_Pets", quote=True
+    )
+    test_column_names: list[str] = column_names_all_uppercase + [
+        quoted_column_name_0,
+        quoted_column_name_1,
+        quoted_column_name_2,
+    ]
+    random.shuffle(test_column_names)
```
Comment (anthonyburdi) [inline tests/expectations/metrics/test_metrics_util.py:435]: Consider deleting this shuffle since we are sorting anyway.

File: tests/expectations/metrics/test_metrics_util.py [hunk 12]
```diff
@@ -272,3 +295,149 @@ def test_get_unexpected_indices_for_multiple_pandas_named_indices_named_unexpect
...
+@pytest.mark.parametrize(
+    "input_column_name, output_column_name, confirm_not_equal_column_name",
+    [
+        [
+            "ARTISTS",
+            "ARTISTS",
+            None,
+        ],
+        [
+            "travel_agents",
```
Comment (anthonyburdi) [inline tests/expectations/metrics/test_metrics_util.py:326]: I'm going to approve, but in the future let's use more descriptive names here. E.g. 
```suggestion
            "should_be_quoted",
```
It really helps with readability.

File: tests/expectations/metrics/test_metrics_util.py [hunk 13]
```diff
@@ -83,6 +89,23 @@ def unexpected_index_list_two_index_columns():
...
+        "engineers",
+        "lawyers",
+        "scientists",
+    ]
+
+
+@pytest.fixture
+def column_names_all_uppercase(column_names_all_lowercase: list[str]) -> list[str]:
+    name: str
+    return [name.upper() for name in column_names_all_lowercase]
```
Comment (anthonyburdi) [inline tests/expectations/metrics/test_metrics_util.py:106]: Consider moving these fixtures down, right above the tests where they are used (e.g. line 298)

File: tests/expectations/metrics/test_metrics_util.py [hunk 14]
```diff
@@ -272,3 +295,149 @@ def test_get_unexpected_indices_for_multiple_pandas_named_indices_named_unexpect
...
+    "input_column_list, output_column_list",
+    [
+        [
+            ["ARTISTS", "travel_agents", "FarmAnimals", "Household_Pets"],
+            [
+                "ARTISTS",
+                sqlalchemy.quoted_name(value="travel_agents", quote=True),
+                sqlalchemy.quoted_name(value="FarmAnimals", quote=True),
+                sqlalchemy.quoted_name(value="Household_Pets", quote=True),
+            ],
```
Comment (anthonyburdi) [inline tests/expectations/metrics/test_metrics_util.py:411]: Consider adding an `id` to each of the params that describes what is being tested. It can help with readability.

**Review by anthonyburdi — APPROVED** [review]
Approving but left some suggestions.

File: tests/expectations/metrics/test_metrics_util.py (hunk 11 above)
Comment (alexsherstinsky) [reply tests/expectations/metrics/test_metrics_util.py:435]: @anthonyburdi This shuffle is needed here in order to intersperse the input column names to insure that the algorithm does not depend on any position.  The sorting at the end is to ensure that the types are correct, regardless of the order.  Happy to discuss.  Thanks.

File: tests/expectations/metrics/test_metrics_util.py (hunk 12 above)
Comment (alexsherstinsky) [reply tests/expectations/metrics/test_metrics_util.py:326]: @anthonyburdi I will make this change in a tiny pull request imminently.  Thanks!

File: tests/expectations/metrics/test_metrics_util.py (hunk 13 above)
Comment (alexsherstinsky) [reply tests/expectations/metrics/test_metrics_util.py:106]: @anthonyburdi I am happy to do that -- in the next pull request.  Should we adopt this as a general policy in test (as in: bunch up fixtures close to where they are used in the module?) -- we should probably discuss pros and cons first.  Thanks!

File: tests/expectations/metrics/test_metrics_util.py (hunk 14 above)
Comment (alexsherstinsky) [reply tests/expectations/metrics/test_metrics_util.py:411]: @anthonyburdi Great idea -- in next pull request coming imminently.  Thanks!

File: tests/expectations/metrics/test_metrics_util.py (hunk 11 above)
Comment (anthonyburdi) [reply tests/expectations/metrics/test_metrics_util.py:435]: Mind adding that as a short comment?

Comment (alexsherstinsky) [reply tests/expectations/metrics/test_metrics_util.py:435]: @anthonyburdi -- absolutely -- in next commit.  Thanks!
